### ISSUE TITLE
Webmin module vs XigmaNAS extension

### ISSUE THREAD

**Issue by tschettervictor** [cbsd/xigmajail#6]
I recently started playing around with Webmin installed on stock FreeBSD. Since XigmaNAS is an appliance OS, would it be more practical to simply write a Webmin module if folks want to have a GUI? 

One would have to know Perl (I don’t) but I think the overall experience might be better.

See https://github.com/JRGTH/bastille-manager-webmin-module for reference.

Thoughts?

**Comment by olevole** [comment]
webmin is another level of abstraction.

And what don't you like about the 'native' framework from XigmaNAS ? It is self-sufficient enough for building forms and is probably supported. If for some reason webmin stops working on XS, it will automatically block the virtual environment management module.

ps: I won't say anything bad about webmin - I used it more than 15 years ago and I liked it ;)

**Comment by tschettervictor** [comment]
I’m ok with XigmaNAS if cbsd can be made to work properly with it (including embedded version)

I was just thinking along the lines of having more control over a server. Webmin gives that control and customization.

You can also install a full XigmaNAS installation which allows you to install additional packages (including Webmin)

Just wanted to know your thoughts. Thanks.

**Comment by olevole** [comment]
In terms of development efforts, it seems there is no difference 

I see that there are pros and cons:

-  already mentioned above, this is an additional layer that is not a basic part of Xigma
+ can be used without Xigma anywhere where webmin works

BTW: I recently discovered **/usr/ports/sysutils/buildah** - It integrates easily with cbsd and can be used to run images as well as generate images. It seems like this could give a semblance of docker to XigmaNAS:
```
bash 
# buildah from --name img docker.io/dougrabson/freebsd14-minimal:latest
Trying to pull docker.io/dougrabson/freebsd15-minimal:latest...
Getting image source signatures
Copying blob 77a9400dd283 done   |
Copying blob 9fd8373ff323 done   |
Copying blob 193f6d37dc31 done   |
Copying blob ceefb323aafc done   |
Copying config 3d2d085d0a done   |
Writing manifest to image destination
img

# cbsd jcreate jname=img interface=0 ip4_addr="172.22.9.14" mnt_start=/tmp/img.mount astart=0
Please wait: this will take a while...
Execute mnt_start script: /tmp/img.mount...
664dc2b178de5de777a970c22ef98a3a7a4766ae331d7bae5e72dbd72faade70
...
jcreate done in 7 seconds

# cbsd jstart img
Execute mnt_start script: /tmp/img.mount...
664dc2b178de5de777a970c22ef98a3a7a4766ae331d7bae5e72dbd72faade70
create epair: epair2:vlan20
set resource limit: [ ]
Starting jail: img, parallel timeout=5
eth0
late_start in progress...
ELF ldconfig path: /lib /usr/lib /usr/lib/compat
32-bit compatibility ldconfig path:
...
Sun Nov 24 00:28:40 CET 2024
jstart done in 3 seconds

# cbsd jlogin img
Custom login command:   /usr/sbin/jexec 82 /usr/bin/env -i -U root sh
img# ps ax
  PID TT  STAT    TIME COMMAND
10570  0  SNJ  0:00.05 sh
10646  0  RN+J 0:00.00 ps ax
img#
```

Can also be used for generation of jail images + `bsd-apps` (marketplace).

**Comment by tschettervictor** [comment]
I was thinking of using _either_ XigmaNAS _or_ vanilla BSD with Webmin on top. I was not suggesting using them together.

Ultimately it’s up to you where you want to focus your development on. If it’s on a Webmin module then it can work on any BSD system. But if it’s on XigmaNAS, then it would require XigmaNAS as the base OS.

**Comment by olevole** [comment]
Most likely, writing a module for webmin will be a different task.

In fact, I have not yet fully studied the XigmaNAS API and what additional benefits it can provide. For example, getting a list of pools

**Comment by tschettervictor** [comment]
Sounds good.

I’m assuming you are doing your tests using the Embedded version?

**Comment by tschettervictor** [comment]
Have you seen Appjail? 
It has pre built jail images that get created and uploaded, and can be deployed in a manner similar to docker-compose using Appjail-director.

I don’t think its feasible to do separate projects. It would be far better if all of us could focus our efforts on one image system.

@DtxdF is very active and helpful.

**Comment by DtxdF** [comment]
> I don’t think it feasible to do separate projects. It would be far better I all of use could focus our efforts on one image system.

CBSD just rocks. It would be great if AppJail could use CBSD images and vice versa, so users of both projects would benefit.

Of course, this is off-topic of this issue.

**Comment by olevole** [comment]
> I’m assuming you are doing your tests using the Embedded version?

Haven't tried it yet. I'd like to stabilize the version for full installation first. I haven't been able to devote enough time to it yet.

**Comment by olevole** [comment]
> Have you seen Appjail? It has pre built jail images that get created and uploaded, and can be deployed in a manner similar to docker-compose using Appjail-director.
> 
> I don’t think it feasible to do separate projects. It would be far better I all of use could focus our efforts on one image system.
> 
> @DtxdF is very active and helpful.

Yes I heard about the Appjail ( however, I haven't had a chance to test it yet ). It is quite problematic to combine Appjail + CBSD projects, but I see no problem in having support for the same image format. By the way, cbsd supports two formats - its own and OCI: CBSD can use official docker/OCI images from docker registry (e.g. docker.io) thanks to `buildah`. So if the Appjail (@DtxdF ?) also supports OCI, this problem is already solved ;-)

**Comment by tschettervictor** [comment]
Maybe @DtxdF could share an image built using appjail with you and see how it imports into CBSD.

**Comment by DtxdF** [comment]
> Maybe @DtxdF could share an image built using appjail with you and see how it imports into CBSD.

I think implementing the OCI standard could help, plus FreeBSD released OCI images, so I think this could be leveraged by jail managers.

**Comment by DtxdF** [comment]
> Maybe @DtxdF could share an image built using appjail with you and see how it imports into CBSD.

But hey @olevole, if you have time, yes, I can share you an image and explain how it works. I think it would be easy to integrate the AppJail image format into your project.

The only problem I see with it is that 30% of the work is done at build-time and the rest at run-time, which has nothing to do with the image format. I think the way AppJail currently generates the InitScript after importing an AppJail image is not portable, but I think we can create a small .subr file to make it portable.

What do you guys think?

**Comment by DtxdF** [comment]
![2024-12-03_23:17:27_1266x510](https://github.com/user-attachments/assets/74452e7b-0fc9-4ac2-b6a4-c5da56e04ced)

Thanks to buildah, I can mount the container file system in the same way as CBSD.

@tschettervictor

**Comment by olevole** [comment]
well, the problem with common images is solved ;)

**Comment by olevole** [comment]
> But hey @olevole, if you have time, yes, I can share you an image and explain how it works. I think it would be easy to integrate the AppJail image format into your project.
> 
> The only problem I see with it is that 30% of the work is done at build-time and the rest at run-time, which has nothing to do with the image format. I think the way AppJail currently generates the InitScript after importing an AppJail image is not portable, but I think we can create a small .subr file to make it portable.
> 
> What do you guys think?

I will try to get to know the Appjail better in the near future to understand your ideas (and as a result, find out whether a third format needs to be added to the CBSD, since the CBSD has such an image format that is already more than 10 years old). 

It should also be taken into account that:

a) CBSD is not only for containers (  I would like to distribute VM images (VM-based services) via OCI + docker registry) ;
b) CBSD is not only for FreeBSD ;-)

I'm currently playing with the OCI images and it seems that focusing on it will be more promising for both projects. 
My notes: https://github.com/cbsd/cbsd/blob/develop/share/docs/general/cbsd_oci.md

I don't want to give up the FreeBSD concept (base in RO + data in RW as overlay) - yes - this is the privilege of FreeBSD images, thanks to a successful hierarchy. Thanks to basejail RO: we have a full set of utilities (for debugging) for free.

At the moment I am thinking about options for combining the FreeBSD concept and OCI images.

**Comment by olevole** [comment]
@DtxdF FYI: keep in mind that [pot](https://www.freshports.org/sysutils/pot) also uses OCI. Not sure about BastilleBSD.
If the AppJail gets OCI support, then shared images are already available for three projects.

**Comment by DtxdF** [comment]
> I will try to get to know the Appjail better in the near future to understand your ideas (and as a result, find out whether a third format needs to be added to the CBSD, since the CBSD has such an image format that is already more than 10 years old).

Thanks @olevole, I will make the changes I mentioned in the InitScripts and let you know when everything is ready.

> I'm currently playing with the OCI images and it seems that focusing on it will be more promising for both projects.

I was surprised by the existence of buildah, and yes, it looks promising for both projects. I'm going to do experiments with it and AppJail. I think we should create an image with a real application. It's not hard since the hard part is done by buildah.

**Comment by tschettervictor** [comment]
Uptime kuma would be a decent start.

**Comment by DtxdF** [comment]
@olevole 

I have implemented OCI in AppJail:

* https://appjail.readthedocs.io/en/latest/OCI/

Now we can create OCI images and use them in both projects!

**Comment by mergar** [comment]
@DtxdF this was the right move) . sorry for the off-topic - do you use entrypoints from OCI?

**Comment by DtxdF** [comment]
> @DtxdF this was the right move) . sorry for the off-topic - do you use entrypoints from OCI?

Thank you!

Yes, it can execute the entrypoint.

**Comment by olevole** [comment]
Right now, honestly, I just don't have the bandwidth to build a Webmin module. Part of the reason is that I don't really have a use case for it myself, and without using it daily, the quality is bound to suffer. Since no one else has stepped up to take this on either, I’m going to close this ticket for now. Hopefully, someone picks it up down the road! Thanks for feedback!
